File: src/0286_walls_and_gates.js
function walls_and_gates(rooms) {
    const rows = rooms.length;
    const cols = rooms[0].length;
    const visit = new Set();
    const q = [];
    let head = 0;
    const addRooms = (r, c) => {
        const key = `${r},${c}`;
        if (Math.min(r, c) < 0 || r === rows || c === cols || visit.has(key) || rooms[r][c] === -1) return;
        visit.add(key);
        q.push([r, c]);
    };
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (rooms[r][c] === 0) {
                q.push([r, c]);
                visit.add(`${r},${c}`);
            }
        }
    }
    let dist = 0;
    while (head < q.length) {
        const levelEnd = q.length;
        for (; head < levelEnd; head++) {
            const [r, c] = q[head];
            rooms[r][c] = dist;
            addRooms(r + 1, c);
            addRooms(r - 1, c);
            addRooms(r, c + 1);
            addRooms(r, c - 1);
        }
        dist++;
    }
}

module.exports = { walls_and_gates };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    let arg0 = [[2147483647, (-1), 0, 2147483647], [2147483647, 2147483647, 2147483647, (-1)], [2147483647, (-1), 2147483647, (-1)], [0, (-1), 2147483647, 2147483647]];
    let _result = walls_and_gates(arg0);
    assert.deepStrictEqual(arg0, [[3, (-1), 0, 1], [2, 2, 1, (-1)], [1, (-1), 2, (-1)], [0, (-1), 3, 4]]);
    arg0 = [[(-1)]];
    _result = walls_and_gates(arg0);
    assert.deepStrictEqual(arg0, [[(-1)]]);
}

File: src/0286_walls_and_gates.rs
struct Solution;

use std::collections::{HashSet, VecDeque};

impl Solution {
    fn add_rooms(
        r: i32,
        c: i32,
        rooms: &[Vec<i32>],
        visit: &mut HashSet<(i32, i32)>,
        q: &mut VecDeque<(i32, i32)>,
    ) {
        let rows = rooms.len() as i32;
        let cols = rooms[0].len() as i32;
        if r.min(c) < 0
            || r == rows
            || c == cols
            || visit.contains(&(r, c))
            || rooms[r as usize][c as usize] == -1
        {
            return;
        }
        visit.insert((r, c));
        q.push_back((r, c));
    }

    pub fn walls_and_gates(rooms: &mut Vec<Vec<i32>>) {
        let rows = rooms.len();
        let cols = rooms[0].len();
        let mut visit = HashSet::new();
        let mut q = VecDeque::new();

        for r in 0..rows {
            for c in 0..cols {
                if rooms[r][c] == 0 {
                    q.push_back((r as i32, c as i32));
                    visit.insert((r as i32, c as i32));
                }
            }
        }

        let mut dist = 0;
        while !q.is_empty() {
            let level_len = q.len();
            for _ in 0..level_len {
                let (r, c) = q.pop_front().unwrap();
                rooms[r as usize][c as usize] = dist;
                Self::add_rooms(r + 1, c, rooms, &mut visit, &mut q);
                Self::add_rooms(r - 1, c, rooms, &mut visit, &mut q);
                Self::add_rooms(r, c + 1, rooms, &mut visit, &mut q);
                Self::add_rooms(r, c - 1, rooms, &mut visit, &mut q);
            }
            dist += 1;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        let mut arg0 = vec![
            vec![2147483647, -1, 0, 2147483647],
            vec![2147483647, 2147483647, 2147483647, -1],
            vec![2147483647, -1, 2147483647, -1],
            vec![0, -1, 2147483647, 2147483647],
        ];
        Solution::walls_and_gates(&mut arg0);
        assert_eq!(
            arg0,
            vec![
                vec![3, -1, 0, 1],
                vec![2, 2, 1, -1],
                vec![1, -1, 2, -1],
                vec![0, -1, 3, 4]
            ]
        );
        let mut arg0 = vec![vec![-1]];
        Solution::walls_and_gates(&mut arg0);
        assert_eq!(arg0, vec![vec![-1]]);
    }
}
